semgrep-core/src/parsing/pfff/json_to_generic.ml

```diff
@@ -65,13 +65,20 @@ let expr x =
                        | Right t -> G.Ellipsis t |> G.e)
                 in
                 G.Container (G.Dict, (lp, zs, rp)) |> G.e
+            | G.L (G.String (s, t)) when unescape_strings ->
+                (* TODO: use the proper JSON unescaping convention, not
+                 * the OCaml one, which is what Scanf.unescape does.
+                 * Reuse Yojson code?
+                 *)
+                G.L (G.String (Scanf.unescaped s, t)) |> G.e
```

Comment:
@mjambon this may require your deeper understanding of JSON than me to do the right thing.

Reply:
oh, I missed this comment earlier. This is incorrect since it evaluates ocaml string literals rather than json ones. We should be able to use yojson's `read_string` function.

Here's code to read a JSON string literal:

```ocaml
let read_json_string json =
  Yojson.Safe.read_string (Yojson.init_lexer ~buf:(Bi_outbuf.create 128) ()) (Lexing.from_string json)
```

```
# read_json_string {|"\"hello\u0000"|};;
- : string = "\"hello\000"
```
It parses the whole string literal including the quotes. There's no public function to parse the only contents. An easy but slightly inefficient way is:

```
let unescape_json_string s =
  read_json_string ("\"" ^ s ^ "\"")
```

Reply:
PR in progress: https://github.com/returntocorp/semgrep/pull/5286